implementations/rust/ockam/ockam_vault/src/secret_impl.rs

```diff
@@ -1,36 +1,38 @@
-use crate::error::Error;
-use crate::software_vault_impl::{SoftwareVaultImpl, VaultEntry};
+use crate::software_vault::{SoftwareVault, VaultEntry};
+use crate::VaultError;
 use arrayref::array_ref;
-use ockam_vault_core::secret::Secret;
-use ockam_vault_core::secret_vault::SecretVault;
-use ockam_vault_core::types::{
-    PublicKey, SecretAttributes, SecretKey, SecretType, CURVE25519_SECRET_LENGTH,
+use ockam_vault_core::{
+    HashVault, PublicKey, Secret, SecretAttributes, SecretKey, SecretType, SecretVault,
+    CURVE25519_SECRET_LENGTH,
 };
 use rand::rngs::OsRng;
 use rand::RngCore;
 use zeroize::Zeroize;
 
-impl SecretVault for SoftwareVaultImpl {
-    fn secret_generate(
-        &mut self,
-        attributes: SecretAttributes,
-    ) -> Result<Secret, ockam_core::Error> {
+impl SecretVault for SoftwareVault {
+    /// Generate fresh secret. Only Curve25519 and Buffer types are supported
+    fn secret_generate(&mut self, attributes: SecretAttributes) -> ockam_core::Result<Secret> {
         let mut rng = OsRng {};
-        let key = match attributes.stype {
+        let (key, key_id) = match attributes.stype {
             SecretType::Curve25519 => {
                 let sk = x25519_dalek::StaticSecret::new(&mut rng);
-                SecretKey::new(sk.to_bytes().to_vec())
+                let public = x25519_dalek::PublicKey::from(&sk);
+                let private = SecretKey::new(sk.to_bytes().to_vec());
+                let key_id = self.sha256(public.as_bytes())?;
```

Comment:
We need to expand all uses of `?` and `unwrap` to `match` or `if let`s throughout the Vault code

Reply:
@jared-s all the unwraps in the vault are inside tests. Except one unwrap in xeddsa @mikelodder7 
And what's wrong with `?`

Reply:
@mikelodder7 fixed the unwrap